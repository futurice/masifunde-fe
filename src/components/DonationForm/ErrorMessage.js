import styled from 'styled-components'
import PropTypes from 'prop-types'

const Error = styled.div`
  color: ${(props) => props.theme.error};
`

const ErrorMessage = ({ meta, ...rest }) =>
  meta.error && meta.touched ? <Error {...rest}>{meta.error}</Error> : null

ErrorMessage.propTypes = {
  meta: PropTypes.shape({
    error: PropTypes.string,
    touched: PropTypes.bool,
  }),
}

ErrorMessage.defaultProps = {
  meta: {},
}

export default ErrorMessage
