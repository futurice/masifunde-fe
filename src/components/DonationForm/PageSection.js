import styled from 'styled-components'
import PageSection from '../PageSection'

const DonationFormPageSection = styled(PageSection)`
  margin-top: 3rem;

  ${({ contained }) => contained && 'padding-right: 0;'}
`

DonationFormPageSection.propTypes = {
  ...PageSection.propTypes,
}

DonationFormPageSection.defaultProps = {
  ...PageSection.defaultProps,
}

export default DonationFormPageSection
