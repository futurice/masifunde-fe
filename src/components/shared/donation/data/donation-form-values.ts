import { CountryCode } from '../../../../i18n/countries'

/**
 * Possible options for what the donation should be used for.
 * Only for the general donation form (not for the campaign ones).
 */
export type FundraisingProjectId =
  | '3520' // Learn 4 Life
  | '3522' // General work in South Africa
  | '3531' // General work in Germany

/**
 * Available options for (not) receiving a donation receipt.
 */
export type ReceiptPreference =
  | 'receipt_now'
  | 'receipt_end_of_year'
  | 'no_receipt'

/**
 * The shape of a donation form's data.
 * Does not include the payment details, which are entered into the
 * Fundraisingbox iframe and are not visible to this website for
 * security reasons.
 */
export type DonationFormValues = {
  // Donation details
  amount?: string // = `value` of a `donationAmount` content entry
  interval?: string // = `value` of a `donationInteval` content entry
  wantsReceipt: ReceiptPreference
  projectId?: FundraisingProjectId
  // Personal details
  country: CountryCode
  email: string
  firstName: string
  lastName: string
  salutation?: 'Mrs.' | 'Mr.'
  address?: string
  city?: string
  postCode?: string
  title?: string
  companyName?: string
}

/**
 * The type of a field name in {@link DonationFormValues}.
 */
export type DonationFormField = keyof DonationFormValues
