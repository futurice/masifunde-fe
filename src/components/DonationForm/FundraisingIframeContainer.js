import PropTypes from 'prop-types'
import { useEffect } from 'react'
import styled from 'styled-components'
import FormContainer from './FormContainer'
import FundraisingIframe from './FundraisingIframe'
import FundraisingboxLink from './FundraisingboxLink'
import PageSection from './PageSection'
import SubHeader from './SubHeader'

/**
 * When scrolling directly to the iframe the top part of the iframe gets
 * hidden behind the fixed header. Scrolling to "ScrollDiv" which is placed
 * further up on the page will display the entire iframe.
 */
const ScrollDiv = styled.div`
  position: relative;
  top: -${({ theme }) => theme.headerHeight};
`

const IFRAME_SCROLL_ID = 'iframe_scroll_anchor'

function FundraisingIframeContainer({
  fields,
  formTitle,
  scrollToIframe,
  ...rest
}) {
  useEffect(() => {
    if (scrollToIframe) {
      window.location.hash = IFRAME_SCROLL_ID
    }
  }, [scrollToIframe])

  return (
    <PageSection contained={false}>
      {!scrollToIframe && (
        <FormContainer>
          <SubHeader>{formTitle}</SubHeader>
        </FormContainer>
      )}
      <ScrollDiv id={IFRAME_SCROLL_ID} />
      <FundraisingIframe {...fields} {...rest} />
      <FormContainer>
        <FundraisingboxLink />
      </FormContainer>
    </PageSection>
  )
}

FundraisingIframeContainer.propTypes = {
  fields: PropTypes.shape(),
  formTitle: PropTypes.string.isRequired,
  scrollToIframe: PropTypes.bool.isRequired,
}

FundraisingIframeContainer.defaultProps = {
  fields: undefined,
}

export default FundraisingIframeContainer
