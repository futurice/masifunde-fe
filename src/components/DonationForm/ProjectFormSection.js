import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'

import { fundraisingFormSpacing } from '../../utils/constants'
import { smBreakpoint } from '../../styling/breakpoints'
import PageSection from './PageSection'
import SubHeader from './SubHeader'
import RadioButton from './RadioButton'
import ErrorMessage from './ErrorMessage'

const CountryLabelContainer = styled.div`
  margin-right: ${fundraisingFormSpacing};
  width: 100%;

  @media (min-width: ${smBreakpoint}) {
    width: calc(50% - ${fundraisingFormSpacing});
  }
`

const ProjectFormSection = ({
  deProjectId,
  fieldName,
  markdownDe,
  markdownSa,
  saProjectId,
  title,
}) => (
  <Field name={fieldName}>
    {({ input, meta }) => (
      <PageSection>
        <SubHeader className="row">{title}</SubHeader>
        <div className="row">
          <CountryLabelContainer data-toggle="buttons">
            <RadioButton
              {...input}
              id="countryInputSa"
              label={markdownSa}
              value={saProjectId}
              isActive={input.value === saProjectId}
            />
          </CountryLabelContainer>
          <CountryLabelContainer>
            <RadioButton
              {...input}
              id="countryInputDe"
              label={markdownDe}
              value={deProjectId}
              isActive={input.value === deProjectId}
            />
          </CountryLabelContainer>
        </div>
        <ErrorMessage meta={meta} />
      </PageSection>
    )}
  </Field>
)

ProjectFormSection.propTypes = {
  deProjectId: PropTypes.string.isRequired,
  fieldName: PropTypes.string.isRequired,
  markdownDe: PropTypes.string.isRequired,
  markdownSa: PropTypes.string.isRequired,
  saProjectId: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
}

export default ProjectFormSection
