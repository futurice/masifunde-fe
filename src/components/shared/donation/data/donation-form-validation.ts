import { Schema, ValidationError, number, string } from 'yup'
import { DonationFormField, DonationFormValues } from './donation-form-values'

// Validations
// ===========

type FieldValidations = Record<DonationFormField, Schema>

/**
 * The field validations that are applied in all cases.
 */
const baseFieldValidations: FieldValidations = {
  amount: number()
    .typeError('donation.mustBeANumber')
    .required('donation.required')
    .integer('donation.mustBeANumber')
    .min(1, 'donation.cantBeLessThan')
    .max(10_000, 'donation.cantBeGreaterThan'),
  interval: string().required('donation.requiredInterval'),
  wantsReceipt: string().required('donation.required'),
  projectId: string(),
  address: string(),
  city: string(),
  postCode: string(),
  country: string().required('donation.required'),
  email: string().required('donation.required').email('donation.invalidEmail'),
  firstName: string().required('donation.required'),
  lastName: string().required('donation.required'),
  salutation: string().required('donation.required'),
  title: string(),
  companyName: string(),
}

/**
 * The field validations that are applied only if the user is
 * expected to choose a project to donate to.
 */
const fieldValidationsIfProjectRequired: Partial<FieldValidations> = {
  projectId: string().required('donation.requiredProject'),
}

/**
 * The field validations that are applied only if the user has
 * requested a donation receipt to be sent to them.
 */
const fieldValidationsIfReceiptRequested: Partial<FieldValidations> = {
  address: string().required('donation.required'),
  city: string().required('donation.required'),
  postCode: string().required('donation.required'),
}

// Types
// =====

/**
 * A validation error for a particular form field.
 */
export type DonationFormFieldError =
  | { key: 'donation.required'; details: undefined }
  | { key: 'donation.requiredProject'; details: undefined }
  | { key: 'donation.requiredInterval'; details: undefined }
  | { key: 'donation.invalidEmail'; details: undefined }
  | { key: 'donation.mustBeANumber'; details: undefined }
  | { key: 'donation.cantBeLessThan'; details: { min: number } }
  | { key: 'donation.cantBeGreaterThan'; details: { max: number } }

/**
 * The return type of {@link validateDonationForm}. For each invalid field,
 * it contains a mapping from the field name to the corresponding error.
 */
export type DonationFormErrors = Record<
  DonationFormField,
  DonationFormFieldError | undefined
>

/**
 * The options that can be passed to {@link validateDonationForm}.
 */
export type DonationFormValidationOptions = {
  /**
   * Whether the user has to select a project to donate to.
   * False by default.
   */
  projectRequired?: boolean
}

// Functions
// =========

/**
 * Checks the current values entered into a donation form according
 * to the given options. Returns a mapping from field names to error
 * messages if at least one field is invalid, or an empty object
 * otherwise.
 *
 * @param values - The donation form data to validate.
 * @param options - Validation options.
 * @returns An object with a name-to-message entry for each invalid field.
 */
export function validateDonationForm(
  values: DonationFormValues,
  options?: DonationFormValidationOptions
): DonationFormErrors {
  const fieldValidations = chooseFieldValidations(values, options)

  const errors: DonationFormErrors = {
    amount: undefined,
    interval: undefined,
    wantsReceipt: undefined,
    projectId: undefined,
    country: undefined,
    email: undefined,
    firstName: undefined,
    lastName: undefined,
    salutation: undefined,
    address: undefined,
    city: undefined,
    postCode: undefined,
    title: undefined,
    companyName: undefined,
  }

  for (const [fieldName, schema] of Object.entries(fieldValidations)) {
    const field = fieldName as DonationFormField
    const value = values[field]
    try {
      schema.validateSync(value)
    } catch (e) {
      errors[field] = toFieldError(e as ValidationError)
    }
  }

  return errors
}

function chooseFieldValidations(
  values: DonationFormValues,
  options?: DonationFormValidationOptions
): FieldValidations {
  let validations = baseFieldValidations

  if (options?.projectRequired) {
    validations = {
      ...validations,
      ...fieldValidationsIfProjectRequired,
    }
  }

  if (values.wantsReceipt !== 'no_receipt') {
    validations = {
      ...validations,
      ...fieldValidationsIfReceiptRequested,
    }
  }

  return validations
}

function toFieldError(e: ValidationError): DonationFormFieldError {
  switch (e.message) {
    case 'donation.cantBeLessThan':
      return {
        key: e.message,
        details: { min: e.params!.min as number },
      }
    case 'donation.cantBeGreaterThan':
      return {
        key: e.message,
        details: { max: e.params!.max as number },
      }
    case 'donation.required':
    case 'donation.requiredProject':
    case 'donation.requiredInterval':
    case 'donation.invalidEmail':
    case 'donation.mustBeANumber':
      return {
        key: e.message,
        details: undefined,
      }
      break
    default:
      throw new Error(`Unexpected validation error: ${e.message}`)
  }
}
