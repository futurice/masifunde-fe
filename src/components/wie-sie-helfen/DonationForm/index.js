import { noop } from 'lodash'
import PropTypes from 'prop-types'
import { Form } from 'react-final-form'
import styled from 'styled-components'
import countries from '../../../utils/countries'
import AmountFormSection from './AmountFormSection'
import FormContainer from './FormContainer'
import FundraisingIframeContainer from './FundraisingIframeContainer'
import InputField from './InputField'
import IntervalFormSection from './IntervalFormSection'
import MultipleInputField from './MultipleInputField'
import PageSection from './PageSection'
import ProjectFormSection from './ProjectFormSection'
import SelectField from './SelectField'
import SubHeader from './SubHeader'
import {
  ADDRESS,
  AMOUNT,
  CITY,
  COMPANY_NAME,
  COUNTRY,
  EMAIL,
  FIRST_NAME,
  LAST_NAME,
  PAYMENT_INTERVAL,
  POST_CODE,
  PROJECT_ID,
  SALUTATION,
  TITLE,
  WANTS_RECEIPT,
} from './constants/fieldNames'
import {
  COUNTRY_GERMANY,
  DE_PROJECT_ID,
  NO_RECEIPT_OPTION_VALUE,
  RECEIPT_NOW_OPTION_VALUE,
  SA_PROJECT_ID,
} from './constants/formValues'
import withFormState from './withFormState'

const HiddenButton = styled.button`
  display: none;
`

const DonationForm = ({
  formTitle,
  fundraisingboxIframeTitle,
  fundraisingboxFormHash,
  validateForm,
  decorators,
  fields,
  projectHeadline,
  buttonProjectDeText,
  buttonProjectSaText,
  amounts,
  intervals,
  otherAmountPlaceholder,
  amountTitle,
  intervalTitle,
  initialValues,
  enableProjectSelection,
  enableOtherAmount,
  minimumYearlyAmount,
  iframeStatus,
  disableIntervalSelection,
}) => {
  let formRef

  const submitForm = () => {
    if (formRef) {
      formRef.click()
    }
  }

  const countriesOptions = Object.entries(countries).map(
    ([countryKey, country]) => ({
      value: countryKey,
      text: country,
    })
  )
  const receiptOptions = [
    { value: RECEIPT_NOW_OPTION_VALUE, text: 'Ja, so schnell wie m√∂glich' },
    {
      value: 'receipt_end_of_year',
      text: 'Ja, konsolidiert am Ende des Jahres',
    },
    {
      value: NO_RECEIPT_OPTION_VALUE,
      text: 'Nein, ich brauche keine Quittung',
    },
  ]
  const salutationOptions = [
    { value: 'Mrs.', text: 'Frau' },
    { value: 'Mr.', text: 'Herr' },
    { value: 'diverse', text: 'Divers' },
  ]
  const isIframeStatusFailureOrSuccess =
    iframeStatus === 'successful' || iframeStatus === 'failure'

  return (
    <>
      {!isIframeStatusFailureOrSuccess && (
        <Form
          onSubmit={noop}
          decorators={decorators}
          initialValues={{
            [WANTS_RECEIPT]: RECEIPT_NOW_OPTION_VALUE,
            [COUNTRY]: COUNTRY_GERMANY,
            // if the interval is disabled then assume the payment will be single
            ...(disableIntervalSelection && { [PAYMENT_INTERVAL]: '0' }),
            ...initialValues,
          }}
          validate={validateForm}
          render={({ handleSubmit, values }) => (
            <form onSubmit={handleSubmit} autoComplete="on">
              <FormContainer>
                {enableProjectSelection && (
                  <ProjectFormSection
                    deProjectId={DE_PROJECT_ID}
                    fieldName={PROJECT_ID}
                    markdownDe={buttonProjectDeText}
                    markdownSa={buttonProjectSaText}
                    saProjectId={SA_PROJECT_ID}
                    title={projectHeadline}
                  />
                )}

                {disableIntervalSelection || (
                  <IntervalFormSection
                    fieldName={PAYMENT_INTERVAL}
                    title={intervalTitle}
                    intervals={intervals}
                  />
                )}

                <AmountFormSection
                  fieldName={AMOUNT}
                  title={amountTitle}
                  amounts={amounts}
                  enableOtherAmount={enableOtherAmount}
                  otherAmountPlaceholder={otherAmountPlaceholder}
                  interval={values[PAYMENT_INTERVAL]}
                  minimumYearlyAmount={minimumYearlyAmount}
                />

                <PageSection contained={false}>
                  <SubHeader>{formTitle}</SubHeader>

                  <SelectField
                    fieldName={SALUTATION}
                    label="Anrede"
                    options={salutationOptions}
                    inputClassName="col-md-4 col-lg-3"
                    autoComplete="honorific-prefix"
                  />

                  <InputField fieldName={TITLE} label="Titel (optional)" />

                  <InputField
                    fieldName={FIRST_NAME}
                    label="Vorname"
                    autoComplete="given-name"
                  />

                  <InputField
                    fieldName={LAST_NAME}
                    label="Nachname"
                    autoComplete="family-name"
                  />

                  <InputField
                    fieldName={EMAIL}
                    label="Email"
                    type="email"
                    autoComplete="email"
                  />

                  <SelectField
                    fieldName={WANTS_RECEIPT}
                    label="Spendenquittung"
                    options={receiptOptions}
                  />

                  {values[WANTS_RECEIPT] !== NO_RECEIPT_OPTION_VALUE && (
                    <>
                      <InputField
                        fieldName={COMPANY_NAME}
                        label="Firma (optional)"
                        autoComplete="organization"
                      />

                      <InputField
                        fieldName={ADDRESS}
                        label="Adresse"
                        autoComplete="street-address"
                      />

                      <MultipleInputField
                        fieldName1={POST_CODE}
                        fieldName2={CITY}
                        autoComplete1="postal-code"
                        autoComplete2="address-level2"
                        label="PLZ / Ort"
                      />
                    </>
                  )}

                  <SelectField
                    fieldName={COUNTRY}
                    label="Land"
                    options={countriesOptions}
                    autoComplete="country"
                  />

                  <HiddenButton
                    ref={(form) => {
                      formRef = form
                    }}
                  >
                    Submit
                  </HiddenButton>
                </PageSection>
              </FormContainer>
            </form>
          )}
        />
      )}

      <FundraisingIframeContainer
        formTitle={fundraisingboxIframeTitle}
        onMouseHover={submitForm}
        hash={fundraisingboxFormHash}
        scrollToIframe={isIframeStatusFailureOrSuccess}
        fields={fields}
      />
    </>
  )
}

DonationForm.propTypes = {
  fundraisingboxIframeTitle: PropTypes.string.isRequired,
  fundraisingboxFormHash: PropTypes.string.isRequired,
  formTitle: PropTypes.string.isRequired,
  fields: PropTypes.shape().isRequired,
  projectHeadline: PropTypes.string,
  buttonProjectDeText: PropTypes.string,
  buttonProjectSaText: PropTypes.string,
  amountTitle: PropTypes.string.isRequired,
  // eslint-disable-next-line react/require-default-props
  amounts: AmountFormSection.propTypes.amounts,
  // eslint-disable-next-line react/require-default-props
  intervals: IntervalFormSection.propTypes.intervals,
  intervalTitle: PropTypes.string,
  otherAmountPlaceholder: PropTypes.string,
  validateForm: PropTypes.func.isRequired,
  initialValues: PropTypes.shape(),
  enableProjectSelection: PropTypes.bool,
  enableOtherAmount: PropTypes.bool,
  decorators: PropTypes.arrayOf(PropTypes.func),
  minimumYearlyAmount: PropTypes.string,
  iframeStatus: PropTypes.oneOf(['successful', 'failure', undefined]),
  disableIntervalSelection: PropTypes.bool,
}

DonationForm.defaultProps = {
  projectHeadline: undefined,
  buttonProjectDeText: undefined,
  buttonProjectSaText: undefined,
  initialValues: {},
  enableProjectSelection: false,
  otherAmountPlaceholder: 'Other',
  enableOtherAmount: false,
  decorators: [],
  minimumYearlyAmount: '0',
  iframeStatus: undefined,
  disableIntervalSelection: false,
  intervals: IntervalFormSection.defaultProps.intervals,
  intervalTitle: '',
}

export default withFormState(DonationForm)
