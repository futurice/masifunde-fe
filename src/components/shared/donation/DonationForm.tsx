import { Decorator } from 'final-form'
import T from 'i18n-react'
import { mapValues, noop, sortBy } from 'lodash'
import { FC, useCallback, useMemo, useRef, useState } from 'react'
import { Form, FormSpy } from 'react-final-form'
import styled from 'styled-components'
import {
  DonationAmount,
  DonationInterval,
} from '../../../content/shared/donation'
import { countryCodes } from '../../../i18n/countries'
import AmountFormSection from './AmountFormSection'
import FormContainer from './FormContainer'
import FundraisingIframeContainer from './FundraisingIframeContainer'
import InputField from './InputField'
import IntervalFormSection from './IntervalFormSection'
import MultipleInputField from './MultipleInputField'
import PageSection from './PageSection'
import ProjectFormSection from './ProjectFormSection'
import SelectField from './SelectField'
import SubHeader from './SubHeader'
import { PAYMENT_INTERVAL, PROJECT_ID } from './constants/fieldNames'
import { DE_PROJECT_ID, SA_PROJECT_ID } from './constants/formValues'
import { validateDonationForm } from './data/donation-form-validation'
import { DonationFormValues } from './data/donation-form-values'

// Props
// =====

export type DonationFormDecorator = Decorator<
  DonationFormValues,
  Partial<DonationFormValues>
>

export type Props = {
  // General
  formTitle: string
  initialValues?: Partial<DonationFormValues>
  decorators?: DonationFormDecorator[]
  // Donation Amount
  amountTitle: string
  amounts: DonationAmount[]
  minimumYearlyAmount?: string
  enableOtherAmount?: boolean
  otherAmountPlaceholder?: string
  // Donation Interval
  intervalTitle: string
  intervals: DonationInterval[]
  disableIntervalSelection?: boolean
  // Project Selection
  enableProjectSelection?: boolean
  projectHeadline?: string
  buttonProjectDeText?: string
  buttonProjectSaText?: string
  // Fundraisingbox
  fundraisingboxIframeTitle: string
  fundraisingboxFormHash: string
  iframeStatus: string | null
}

// Helpers
// =======

const HiddenSubmitButton = styled.button`
  display: none;
`

const receiptOptions = [
  {
    value: 'receipt_now',
    text: 'Ja, so schnell wie mÃ¶glich',
  },
  {
    value: 'receipt_end_of_year',
    text: 'Ja, konsolidiert am Ende des Jahres',
  },
  {
    value: 'no_receipt',
    text: 'Nein, ich brauche keine Quittung',
  },
]

const salutationOptions = [
  { value: 'Mrs.', text: 'Frau' },
  { value: 'Mr.', text: 'Herr' },
]

function getCountryOptions(locale: string) {
  const displayNames = new Intl.DisplayNames([locale, 'en'], { type: 'region' })

  const options = countryCodes.map((code) => ({
    value: code,
    text: displayNames.of(code) ?? code,
  }))

  return sortBy(options, (o) => o.text)
}

function isFundraisingboxIframeFinished(iframeStatus: string | null) {
  return iframeStatus === 'successful' || iframeStatus === 'failure'
}

// Components
// ==========

const DonationForm: FC<Props> = ({
  formTitle,
  fundraisingboxIframeTitle,
  fundraisingboxFormHash,
  decorators,
  projectHeadline,
  buttonProjectDeText,
  buttonProjectSaText,
  amounts,
  intervals,
  otherAmountPlaceholder,
  amountTitle,
  intervalTitle,
  initialValues,
  enableProjectSelection,
  enableOtherAmount,
  minimumYearlyAmount = '0',
  iframeStatus,
  disableIntervalSelection,
}) => {
  const fullInitialValues = useMemo<DonationFormValues>(
    () => ({
      wantsReceipt: 'receipt_now',
      country: 'DE',
      email: '',
      firstName: '',
      lastName: '',
      ...(disableIntervalSelection && { interval: '0' }),
      ...initialValues,
    }),
    [initialValues, disableIntervalSelection]
  )

  const [values, setValues] = useState<DonationFormValues>(fullInitialValues)
  const hiddenSubmitButtonRef = useRef<HTMLButtonElement>(null)

  const countryOptions = getCountryOptions('de')
  const isSubmissionComplete = isFundraisingboxIframeFinished(iframeStatus)

  const validateForm = useCallback(
    (values: DonationFormValues) => {
      const errors = validateDonationForm(values, {
        projectRequired: enableProjectSelection,
      })
      return mapValues(errors, (error) => {
        return error
          ? (T.translate(error.key, error.details) as string)
          : undefined
      })
    },
    [enableProjectSelection]
  )

  const forceValidation = () => {
    // Force react-final-form to validate all form values
    // by clicking a hidden submit button in the form.
    // This is to make the user aware of missing / invalid
    // data as they move to the Fundraisingbox iframe
    // (which displays the payment data section).
    hiddenSubmitButtonRef.current?.click()
  }

  return (
    <>
      {!isSubmissionComplete && (
        <Form<DonationFormValues>
          onSubmit={noop}
          decorators={decorators}
          initialValues={values}
          validate={validateForm}
          render={({ handleSubmit, values }) => (
            <>
              <FormSpy<DonationFormValues>
                subscription={{ values: true, valid: true }}
                onChange={(state) => {
                  if (state.valid) {
                    setValues(state.values)
                  }
                }}
              />

              <form onSubmit={handleSubmit} autoComplete="on">
                <FormContainer>
                  {enableProjectSelection && (
                    <ProjectFormSection
                      deProjectId={DE_PROJECT_ID}
                      fieldName={PROJECT_ID}
                      markdownDe={buttonProjectDeText ?? ''}
                      markdownSa={buttonProjectSaText ?? ''}
                      saProjectId={SA_PROJECT_ID}
                      title={projectHeadline ?? ''}
                    />
                  )}

                  {disableIntervalSelection || (
                    <IntervalFormSection
                      fieldName="interval"
                      title={intervalTitle}
                      intervals={intervals}
                    />
                  )}

                  <AmountFormSection
                    fieldName="amount"
                    title={amountTitle}
                    amounts={amounts}
                    enableOtherAmount={enableOtherAmount}
                    otherAmountPlaceholder={otherAmountPlaceholder}
                    interval={values[PAYMENT_INTERVAL]}
                    minimumYearlyAmount={minimumYearlyAmount}
                  />

                  <PageSection contained={false}>
                    <SubHeader>{formTitle}</SubHeader>

                    <SelectField
                      fieldName="salutation"
                      label="Anrede"
                      options={salutationOptions}
                      inputClassName="col-md-4 col-lg-3"
                      autoComplete="honorific-prefix"
                    />

                    <InputField fieldName="title" label="Titel (optional)" />

                    <InputField
                      fieldName="firstName"
                      label="Vorname"
                      autoComplete="given-name"
                    />

                    <InputField
                      fieldName="lastName"
                      label="Nachname"
                      autoComplete="family-name"
                    />

                    <InputField
                      fieldName="email"
                      label="Email"
                      type="email"
                      autoComplete="email"
                    />

                    <SelectField
                      fieldName="wantsReceipt"
                      label="Spendenquittung"
                      options={receiptOptions}
                    />

                    {values.wantsReceipt !== 'no_receipt' && (
                      <>
                        <InputField
                          fieldName="companyName"
                          label="Firma (optional)"
                          autoComplete="organization"
                        />

                        <InputField
                          fieldName="address"
                          label="Adresse"
                          autoComplete="street-address"
                        />

                        <MultipleInputField
                          fieldName1="postCode"
                          fieldName2="city"
                          autoComplete1="postal-code"
                          autoComplete2="address-level2"
                          label="PLZ / Ort"
                        />
                      </>
                    )}

                    <SelectField
                      fieldName="country"
                      label="Land"
                      options={countryOptions}
                      autoComplete="country"
                    />

                    <HiddenSubmitButton ref={hiddenSubmitButtonRef} />
                  </PageSection>
                </FormContainer>
              </form>
            </>
          )}
        />
      )}

      <FundraisingIframeContainer
        formTitle={fundraisingboxIframeTitle}
        onMouseHover={forceValidation}
        hash={fundraisingboxFormHash}
        scrollToIframe={isSubmissionComplete}
        fields={values}
      />
    </>
  )
}

export default DonationForm
