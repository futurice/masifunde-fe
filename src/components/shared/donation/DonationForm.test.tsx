import { render } from '@testing-library/react'
import DonationForm from './DonationForm'
import { DonationFormValues } from './data/donation-form-values'

test('renders', () => {
  const amounts = [
    { value: 10, text: '10€' },
    { value: 20, text: '20€' },
    { value: 30, text: '30€' },
  ]

  const intervals = [
    { name: 'Once', value: '0' },
    { name: 'Monthly', value: '1' },
    { name: 'Yearly', value: '2' },
  ]

  const initialValues: Partial<DonationFormValues> = {
    projectId: '3522',
  }

  const { getByText } = render(
    <DonationForm
      formTitle={'What would you like to donate for?'}
      projectHeadline={'What would you like to donate for?'}
      buttonProjectDeText={'Projects in Germany'}
      buttonProjectSaText={'Projects in South Africa'}
      intervalTitle={'How often would you like to donate?'}
      intervals={intervals}
      amountTitle="How much would you like to donate?"
      amounts={amounts}
      otherAmountPlaceholder={'Other Amount'}
      fundraisingboxIframeTitle={'Payment Method'}
      fundraisingboxFormHash="abc123"
      initialValues={initialValues}
      enableProjectSelection
      enableOtherAmount
      iframeStatus={null}
    />
  )

  for (const amount of amounts) {
    expect(getByText(amount.text)).toMatchSnapshot()
  }
})
