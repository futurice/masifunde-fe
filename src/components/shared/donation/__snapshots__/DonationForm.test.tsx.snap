// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<label
  class="sc-47980618-1-label sc-41983549-0 KISAm bBQhoO btn"
  for="amountInputOption10"
>
  10€
</label>
`;

exports[`renders 2`] = `
<label
  class="sc-47980618-1-label sc-41983549-0 KISAm bBQhoO btn"
  for="amountInputOption20"
>
  20€
</label>
`;

exports[`renders 3`] = `
<label
  class="sc-47980618-1-label sc-41983549-0 KISAm bBQhoO btn"
  for="amountInputOption30"
>
  30€
</label>
`;
